### PR TITLE
First step to the test suite

### LINKED ISSUE #71: Test suite structure
Since we badly need tests, I suggest we discuss how to organize them so we can start working on them ASAP. I can think of two possible options with two suboptions each.

From my perspective, the main issue with hierarchical structures is that importing classes, functions, etc., might result in very long paths.

#### Option 1a—flat

```
test/
├─ backend.js
├─ format.js
├─ util.js
├─ ...
├─ element.js
├─ local.js
├─ remote.js
├─ coda-api.js
├─ coda-table.js
├─ ...
```

#### Option 1b—hierarchical

```
test/
├─ backends/
│  ├─ basic/
│  │  ├─ element.js
│  │  ├─ local.js
│  │  ├─ remote.js
│  ├─ coda/
│  │  ├─ coda-api.js
│  │  ├─ coda-table.js
│  ├─ .../
├─ backend.js
├─ format.js
├─ util.js
├─ ...
```

#### Option 2a—flat

```
backends/
├─ basic/
│  ├─ element.js
│  ├─ local.js
│  ├─ remote.js
│  ├─ ...
│  ├─ test/
│  │  ├─ element.js
│  │  ├─ local.js
│  │  ├─ remote.js
│  │  ├─ index-fn.js (all tests for the Basic backends)
├─ coda/
│  ├─ api/
│  ├─ table/
│  ├─ ...
│  ├─ test/
│  │  ├─ coda-api.js
│  │  ├─ coda-table.js
│  │  ├─ index-fn.js (all tests for the Coda backends)
├─ .../
src/
├─ test/
│  ├─ backend.js
│  ├─ format.js
│  ├─ util.js
│  ├─ ...
│  ├─ index-fn.js
```

#### Option 2b—hierarchical

```
backends/
├─ basic/
│  ├─ element.js
│  ├─ local.js
│  ├─ remote.js
│  ├─ ...
│  ├─ test/
│  │  ├─ element.js
│  │  ├─ local.js
│  │  ├─ remote.js
│  │  ├─ index-fn.js (all tests for the Basic backends)
├─ coda/
│  ├─ api/
│  ├─ table/
│  ├─ ...
│  ├─ test/
│  │  ├─ api/
│  │  │  ├─ index.js (or coda-api.js ?)
│  │  ├─ table/
│  │  │  ├─ index.js (or coda-table.js ?)
│  │  ├─ index-fn.js (all tests for the Coda backends)
├─ github/
│  ├─ api/
│  ├─ file/
│  ├─ gist/
│  ├─ ...
│  ├─ test/
│  │  ├─ api/
│  │  │  ├─ index.js (or github-api.js ?)
│  │  ├─ file/
│  │  │  ├─ index.js (or github-file.js ?)
│  │  ├─ gist/
│  │  │  ├─ index.js (or github-gist.js ?)
│  │  ├─ index-fn.js (all tests for the GitHub backends)
├─ .../
src/
├─ test/
│  ├─ backend.js
│  ├─ format.js
│  ├─ util.js
│  ├─ ...
│  ├─ index-fn.js
```

Thoughts? Any other suggestions?

### PR DESCRIPTION
**Pull request by DmitrySharabin**
- Add basic test suite structure, closes #71
- Add tests for regressions before merging #70

### CODE REVIEW

**Comment by DmitrySharabin** [comment]
We'll get the following picture after merging the URLPattern PR—no regressions! 🥳

**P. S.** Empty rows correspond to Dropbox and CodaAPI backend tests. There is nothing to test yet (no `ref` object different from the base one with the `url` field).

SCR-20240316-taov

**Review by LeaVerou — APPROVED** [review]
Looks great, thank you so much! See comment wrt Firebase. 

How was the experience of using hTest?

File: backends/firebase/test.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Firebase doesn't work in node.
+// Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data, and node are supported by the default ESM loader.)
```
[the remark this answers is not in the document]
Comment (DmitrySharabin) [reply backends/firebase/test.js:2]: Unfortunately, it does. I also hoped to test at least static methods, but I couldn't manage that yet. I will do my best to find a solution (or a workaround) ASAP. I might not see it now (or do something wrong). We'll see. :)

**Comment by DmitrySharabin** [comment]
> How was the experience of using hTest?

It went so smoothly. Finding a good name and a description for a group of tests was more challenging than writing the tests themselves! 🙃 Inheritance of some main test properties is something—it saved me so many keystrokes. I also worked with the `data` object in the first iteration of tests—it was a breeze to pass data to the run function and change its behavior depending on the context. In my humble opinion, hTest is a highly well-designed product. I bow to you!

File: backends/firebase/test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (LeaVerou) [reply backends/firebase/test.js:2]: K, don't worry about it for now. :)

**Pull request merged**